### PR TITLE
Closes #7 - do not raise exceptions and implement String.Chars protocol

### LINKED ISSUE #7: Breaking changes coming to Bamboo v.2.0.0
👋    Hi Parker,

Thanks for maintaining this Bamboo adapter. 🥳 

There will be a breaking change coming to Bamboo in v2.0.0 where the `deliver_now` function will no longer raise an error on failure to deliver an email. Instead, it will return `:ok` or `:error` tuples. That way, users of the library can decide how to handle the errors. I've completed that work in https://github.com/thoughtbot/bamboo/pull/571.

I've also added a `deliver_now!` function that is meant to behave exactly as `deliver_now` currently behaves. So, there's an easy upgrade path for users who don't want to handle their own errors. They can change `deliver_now` -> `deliver_now!`.

(There are also `deliver_later` corresponding changes, but from the adapter's perspective, I don't know that it matters).

## What that means for adapters

In order for people to upgrade to Bamboo v2.0.0, the adapters will need to stop raising errors when they fail to deliver emails. In order to accommodate that, the adapter's callbacks are changing like this:

```diff
- @callback deliver(%Bamboo.Email{}, %{}) :: any
+ @callback deliver(%Bamboo.Email{}, %{}) :: {:ok, any} | {:error, Exception.t() | String.t()}
```

In other words, adapters should now return an `{:ok, response}` tuple or an `{:error, error}` tuple, where the error is either an exception struct that can later be raised or an error message.

To facilitate that work, [PR 571] also introduces a `Bamboo.ApiError.build_api_error` function that can be used by adapters &mdash; if you want to easily build an error to return in the `{:error, error}` response.

### Sample changes to other adapters

I have already done some changes to the Sendgrid, Mailgun, and Mandrill adapters. They might serve as a guide:

- Sendgrid changes: https://github.com/thoughtbot/bamboo/pull/572
- Mailgun changes: https://github.com/thoughtbot/bamboo/pull/573
- Mandrill changes: https://github.com/thoughtbot/bamboo/pull/574

## Why open this issue?

I wanted to open this issue for two reasons:

1. To let you know about the changes that are coming up (all of the above), and
2. To ask you if there's anything I can do to help with the transition. 

Unfortunately, I can't offer to open PRs to make the changes in all the adapter repos, but if it's helpful, I'd be happy to review PRs.

And if there's some change to Bamboo that would make it easier to work with your adapter, or if you have suggestions or comments on [PR 571], I'd love to know.

Thanks again for maintaining this adapter, and please feel free to close this issue whenever you've read it (if it's of no further use). It just seemed like the easiest way to communicate about the upcoming changes with you.

[PR 571]: https://github.com/thoughtbot/bamboo/pull/571

### PR DESCRIPTION
**Pull request by lovebes**
I also updated the README because of my fresh experience implementing it in Google Workspace (eck not quite rolling off the tongue like GSuite eh? ;)

### CODE REVIEW

File: mix.exs [hunk 1]
```diff
@@ -27,8 +27,8 @@ defmodule BambooGmail.MixProject do
   # Run "mix help deps" to learn about dependencies.
   defp deps do
     [
-      {:bamboo, "~> 1.3"},
-      {:goth, "~> 1.1.0"},
+      {:bamboo, "~> 2.1"},
```
Comment (lovebes) [inline mix.exs:30]: not quite sure how you want to tie the versions together. I'm not sure if the error handling will break for users of bamboo versions ~> 1.3

**Comment by lovebes** [comment]
Actually I think there might be more work. I'll go through it more - will let you know when it is ready for review.

I get this error and maybe it's realted to version upgrade for bamboo:

`** (CaseClauseError) no case clause matching: %HTTPoison.Response{body: "{\n  \"id\": \"17b47acf31eeba00\",\n  \"threadId\": \"17b47acf31eeba00\",\n  \"labelIds\": [\n    \"SENT\"\n  ]\n}\n", headers: [{"Content-Type", "application/json; charset=UTF-8"}, {"Vary", "X-Origin"}, {"Vary", "Referer"}, {"Date", "Sun, 15 Aug 2021 02:38:42 GMT"}, {"Server", "ESF"}, {"Cache-Control", "private"}, {"X-XSS-Protection", "0"}, {"X-Frame-Options", "SAMEORIGIN"}, {"X-Content-Type-Options", "nosniff"}, {"Alt-Svc", "h3=\":443\"; ma=2592000,h3-29=\":443\"; ma=2592000,h3-T051=\":443\"; ma=2592000,h3-Q050=\":443\"; ma=2592000....`
in 

` (bamboo 2.1.0) lib/bamboo/mailer.ex:208: Bamboo.Mailer.deliver_now/4
    (bamboo 2.1.0) lib/bamboo/mailer.ex:231: Bamboo.Mailer.deliver_now!/4`

**Comment by lovebes** [comment]
okay now it's ready. However I am not quite sure how I can write the test to cover my changes for `deliver` - just how would you be mocking the Google API calls? The right way to do it would be to use Mox but that would probably have to have changes in `config` files. Wondering if I should touch more of the code... hm

What do you think @parkerduckworth ?

**Comment by lovebes** [comment]
this closes #7

File: lib/bamboo/adapters/gmail_adapter.ex [hunk 2]
```diff
@@ -93,8 +93,8 @@ defmodule Bamboo.GmailAdapter do
     |> put_bcc(email)
     |> put_from(email)
     |> put_subject(email)
-    |> put_html_body(email)
     |> put_text_body(email)
+    |> put_html_body(email)
```
Comment (parkerduckworth) [inline lib/bamboo/adapters/gmail_adapter.ex:97]: 👍

File: README.md [hunk 3]
```diff
@@ -25,10 +25,17 @@ This adapter allows GSuite account holders to use Bamboo in light of this deprec
...
 2. Implement [server-side authorization](https://developers.google.com/gmail/api/auth/web-server) using Service Account credentials
+  - Don't give any GCP roles for the service account.
+  - Add your Google Workspace email as a user, as servicer account Owner role.
+  - If you have other Domain aliases, add those alias users in https://mail.google.com/mail/u/2/#settings/accounts, in "Send mail as"
 3. Grant the service account domain-wide authority
-4. Authorize API client with required scopes
+  - first do here: https://console.cloud.google.com/iam-admin/serviceaccounts, by clicking the service account, and going to Domain Wide Delegation
+  - Service account will create a related OAuth2.0 client id. 
+4. Authorize API client with required scopes here: https://admin.google.com/u/2/ac/owl/domainwidedelegation
+  - The scope that you need to add is: `https://www.googleapis.com/auth/gmail.send`
```
Comment (parkerduckworth) [inline README.md:38]: Much appreciated 👍

File: mix.exs (hunk 1 above)
Comment (parkerduckworth) [reply mix.exs:30]: I'll tag specific versions. This is fine for now.

**Review by parkerduckworth — APPROVED** [review]
@lovebes thanks for your contribution!!!

**Comment by parkerduckworth** [comment]
> okay now it's ready. However I am not quite sure how I can write the test to cover my changes for `deliver` - just how would you be mocking the Google API calls? The right way to do it would be to use [Mox](https://hexdocs.pm/mox/Mox.html) but that would probably have to have changes in `config` files. Wondering if I should touch more of the code... hm
> 
> What do you think @parkerduckworth ?

Let's open a separate issue to address testing, so we can keep all the discussion in one place.

**Pull request merged**